Reader: make header parsing more robust

chrisnovakovic opened the pull request:
Simplify the header parsing operation in `Reader` by condensing the mutually-dependent parsing functions into the `Next` function - it's longer now, but the logic is easier to follow. Also avoid ignoring what ought to be fatal failure cases by bubbling up non-nil errors from function calls to the caller.

The restructured `Next` function correctly parses GNU and BSD ar headers - previously the trailing slash inserted into file name headers by GNU ar was not always correctly stripped. (The old function structure made it hard to tell that this wouldn't always happen.)

The new test files contain both short and long file names, and are more representative of real-world archive files than the others in this repo. They were generated with llvm-ar 15 as follows:

```
perl -e 'foreach my $i (1 .. 20) { open $f, ">", ($i . ("x" x ($i - length $i))); print $f "The name of this file contains $i character(s).\n" }'
llvm-ar-15 --format=gnu r fixtures/long_filenames_gnu.a 1 2x 3xx 4xxx 5xxxx 6xxxxx 7xxxxxx 8xxxxxxx 9xxxxxxxx 10xxxxxxxx 11xxxxxxxxx 12xxxxxxxxxx 13xxxxxxxxxxx 14xxxxxxxxxxxx 15xxxxxxxxxxxxx 16xxxxxxxxxxxxxx 17xxxxxxxxxxxxxxx 18xxxxxxxxxxxxxxxx 19xxxxxxxxxxxxxxxxx 20xxxxxxxxxxxxxxxxxx
llvm-ar-15 --format=bsd r fixtures/long_filenames_bsd.a 1 2x 3xx 4xxx 5xxxx 6xxxxx 7xxxxxx 8xxxxxxx 9xxxxxxxx 10xxxxxxxx 11xxxxxxxxx 12xxxxxxxxxx 13xxxxxxxxxxx 14xxxxxxxxxxxx 15xxxxxxxxxxxxx 16xxxxxxxxxxxxxx 17xxxxxxxxxxxxxxx 18xxxxxxxxxxxxxxxx 19xxxxxxxxxxxxxxxxx 20xxxxxxxxxxxxxxxxxx
```

toastwaffle commented on reader.go:
```diff
@@ -63,10 +72,21 @@ var (
+	r io.Reader
+
+	// nb is the number of bytes in the current data section that remain unread.
+	nb int64
```
How do you feel about renaming this to bytesRemaining?

toastwaffle commented on reader.go:
```diff
@@ -128,21 +148,29 @@ func (rd *Reader) skipUnread() error {
+		Gid:     int(rd.numeric(s.next(6))),
+		Mode:    rd.octal(s.next(8)),
+		Size:    rd.numeric(s.next(10)),
+	}
```
Just to check: Does Go execute this line by line, or in field definition order? Is it worth a comment saying that changing the order will make it explode?

toastwaffle commented on reader.go:
```diff
@@ -151,65 +179,67 @@ func (rd *Reader) readHeader() (*Header, error) {
+	// The ar format only supports file names of up to 16 bytes to be stored in the file name header
+	// field. Various archivers have adopted different conventions for storing files with longer
+	// names:
+	// - GNU ar replaces the file name in the header with the character "/" followed by an integer. The
```
Given this list format, I'd expect more than one entry?

toastwaffle replied:
Oh, I see the other entry is further down. That's mildly confusing. At the very least I think this "GNU ar" entry should go with the `else if header.Name[0] == '/'`?

toastwaffle commented on reader.go:
```diff
@@ -151,65 +179,67 @@ func (rd *Reader) readHeader() (*Header, error) {
+		if err != nil {
+			return nil, fmt.Errorf("ar: invalid GNU long file name offset '%s'", header.Name[1:])
 		}
+		fileName := rd.stringTable[start:]
```
You mention that it is convention that the string table is the first file in the archive; should we validate that we have a stringTable here instead of potentially exploding?

chrisnovakovic replied:
I was actually wrong about this - the symbol table, if there is one, will always be first. Nevertheless, you're right: we shouldn't assume a symbol table has been defined. I'll fix this.

toastwaffle commented on reader.go:
```diff
@@ -151,65 +179,67 @@ func (rd *Reader) readHeader() (*Header, error) {
+		header.Name = string(bytes.TrimRight(b, " \x00"))
+	// Otherwise, the file name fits in the standard header field.
+	} else {
+		// GNU ar appends "/" to file names stored in the header field. We don't necessarily know whether
```
This comment (or at least part of it) probably should come before line 212

toastwaffle commented on reader.go:
```diff
@@ -151,65 +179,67 @@ func (rd *Reader) readHeader() (*Header, error) {
+			return nil, fmt.Errorf("ar: invalid GNU long file name offset '%s'", header.Name[1:])
 		}
+		fileName := rd.stringTable[start:]
+		end := bytes.IndexByte(fileName, '\n')
```
Do we need to handle the possibility of `\n` being a character in the actual filename?

toastwaffle commented on reader_test.go:
```diff
@@ -123,30 +126,36 @@ func TestReadMulti(t *testing.T) {
+					assert.Equal(t, fmt.Sprintf("%d%s", i, strings.Repeat("x", i - len(strconv.Itoa(i)))), hdr.Name)
+					io.Copy(&buf, reader)
+					expected := []byte(fmt.Sprintf("The name of this file contains %d character(s).\n", i))
+					assert.EqualValues(t, hdr.Size, len(expected))
```
Why `EqualValues` here and not just `Equal`?

chrisnovakovic commented:
Thanks for the review - [d09ecfe](https://github.com/please-build/ar/pull/4/commits/d09ecfeb666fc8aeaff8dd7e8d76b1c1b2aa4eab) addresses the comments and makes parsing more robust. In particular, it detects which variant of the file format the archive uses when the `Reader` is created and only allows file names permitted by that variant when each header is parsed.

toastwaffle commented on reader_test.go:
```diff
@@ -123,30 +126,36 @@ func TestReadMulti(t *testing.T) {
 	}{
-		{"BSD format", "./fixtures/bsd_long_filename.a"},
-		{"GNU format", "./fixtures/gnu_long_filename.a"},
+		{"BSD format", "./fixtures/long_filenames_bsd.a"},
```
Please could you document how these were generated?

toastwaffle replied:
Ah, my bad. Clearly I read that and then forgot all about it.
